prow/cmd/hook/main.go

```diff
@@ -193,41 +194,45 @@ func main() {
 		BugzillaClient:            bugzillaClient,
 	}
 
-	promMetrics := hook.NewMetrics()
+	promMetrics := servemux.NewMetrics()
 
 	defer interrupts.WaitForGracefulShutdown()
 
 	// Expose prometheus metrics
 	metrics.ExposeMetrics("hook", configAgent.Config().PushGateway)
 	pjutil.ServePProf()
 
-	server := &hook.Server{
+	var server servemux.EventHandler
+	server = &hook.Server{
 		ClientAgent:    clientAgent,
 		ConfigAgent:    configAgent,
 		Plugins:        pluginAgent,
 		Metrics:        promMetrics,
 		TokenGenerator: secretAgent.GetTokenGenerator(o.webhookSecretFile),
 	}
+
+	logger := logrus.NewEntry(logrus.New())
+	handler := servemux.NewServeMuxHandler(server, secretAgent.GetTokenGenerator(o.webhookSecretFile), logger)
+
 	interrupts.OnInterrupt(func() {
-		server.GracefulShutdown()
+		handler.GracefulShutdown()
 		if err := gitClient.Clean(); err != nil {
 			logrus.WithError(err).Error("Could not clean up git client cache.")
 		}
 	})
 
-	health := pjutil.NewHealth()
+	mux := servemux.InitializeServeMux(logger, "/hook", handler, nil)
```

Comment:
This is still messy. Why do I need to create a new handler, then initialize it with itself and a `/hook` string?

I want to see a library that takes an exported `Options` struct (which knows how to add itself to a `FlagSet`) and does _all_ the initalization with it. The `main()` of `hook` and `needs-rebase` for instance should be 100% identical in these regards. Adding a new external plugin should consist of:

1. importing and using the `Options`
2. adding on external items
3. registering plugins